openff/interchange/components/smirnoff.py

```diff
@@ -1521,17 +1584,55 @@ def store_matches(
                         duplicate_molecule_particle
                     )
 
-                    for mult in match_mults[(unique_molecule_particle_index,)]:
-                        topology_key = TopologyKey(
-                            atom_indices=(topology_particle_index,),
-                            mult=mult,
-                        )
-                        reference_key = TopologyKey(
-                            atom_indices=(unique_molecule_particle_index,), mult=mult
-                        )
-                        potential_key = matches[reference_key]
+                    # Copy the keys associated with the reference molecule to the duplicate molecule
+                    for key in matches:
+                        if key.this_atom_index == unique_molecule_particle_index:
+                            new_key = key.__class__(**key.dict())
+                            new_key.this_atom_index = topology_particle_index
+
+                            # Have this new key (on a duplicate molecule) point to the same potential
+                            # as the old key (on a unique/reference molecule)
+                            self.slot_map[new_key] = matches[key]
+
+                    # for key in _slow_key_lookup_by_atom_index(
+                    #     matches,
+                    #     topology_particle_index,
+                    # ):
+                    #     self.slot_map[key] = matches[key]
+
+        topology_charges = [0.0] * topology.n_atoms
+        for key, val in self.get_charges().items():
+            topology_charges[key.atom_indices[0]] = val.m
+        # charges: List[float] = [v.m for v in self.get_charges().values()]
+
+        # TODO: Better data structures in Topology.identical_molecule_groups will make this
+        #       cleaner and possibly more performant
+        for molecule in topology.molecules:
+            molecule_charges = [0.0] * molecule.n_atoms
+
+            for atom in molecule.atoms:
+                molecule_index = molecule.atom_index(atom)
+                topology_index = topology.atom_index(atom)
+
+                molecule_charges[molecule_index] = topology_charges[topology_index]
+
+            charge_sum = sum(molecule_charges)
+            formal_sum = molecule.total_charge.m
 
-                        self.slot_map[topology_key] = potential_key
+            if abs(charge_sum - formal_sum) > 0.01:
```

Comment:
To help our memory, this value was selected to match long-standing behavior in the toolkit:

https://github.com/openforcefield/openff-toolkit/blob/fade767977cda3c2d70399ac38644aa5428414fe/openff/toolkit/typing/engines/smirnoff/parameters.py#L3994-L3997